### ISSUE TITLE
sql: support ordering by JSON

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#35706]
Currently it's not possible to order a table by a JSON column.

This is because JSON does not have a valid lexicographic ordering.

for reference, PostgreSQL supports it.

**Comment by knz** [comment]
@rohany @RaduBerinde I think the recent key encoding for arrays will solve this issue?

**Comment by rohany** [comment]
No, the key encoding for arrays will just be for arrays -- i think json ordering is a different beast.

**Comment by awoods187** [comment]
I'm putting this back into triage based on how high it appears in the unimplemented query list cc @rytaft @mgartner

**Comment by mgartner** [comment]
@awoods187 have we heard directly from customers that this is a strongly desired feature? I can't think of a legitimate case where ordering by JSON would be useful or required. Perhaps some ORMs are doing it for _reasons_.

**Comment by knz** [comment]
one reason i could imagine would be to make the output deterministic in tests

**Comment by cucaroach** [comment]
@awoods187 do we have any specific customer examples or motivations for including in this release?   It will be a non-trivial amount of work.  Thanks!

**Comment by awoods187** [comment]
I'll respond in Slack

**Comment by saurik** [comment]
@mgartner FWIW, I ended up deciding to not use cockroach--so I am not myself a customer--but I subscribed to this bug a long time ago, and I'd say you might have the motivating examples in the linked issues #24436 and #35260 (which were closed marks by this one as a duplicate); like, I think this issue might be framing it in your head as a case where the exact *ordering* somehow mattered, but I think a lot of the people who were caring about this actually just wanted uniqueness to work (though in the case of the UNION I wonder if they would have been OK with a UNION ALL; the DISTINCT, though, to me, sounds inherently reasonable). (I think I actually myself *did* have a use case where I wanted the ordering to explicitly work, but, again: I am not a customer/user as, after careful evaluation, I looked elsewhere and gave up on cockroach.)

**Comment by mgartner** [comment]
Thanks for the additional perspective, @saurik. I agree that allowing users to create a uniqueness constraint (backed by a unique index) on a JSON column is a strong motivator for this.
